### PR TITLE
Add progress bar for compilations tasks

### PR DESCRIPTION
**Pull request by KochynVolodymyr**
> ### Motivate of the pull request
> - [ ] To address an existing issue. If so, please provide a link to the issue: <issue id>
> - [ ] Breaking new feature. If so, please describe details in the description part.

> ### Describe the technical details
> #### What is currently done? (Provide issue link if applicable)
> 
> 
> 
> 
> 
>
> #### What does this pull request change?
> 
> 
> 
> 
> 

> ### Which part of the code base require a change
> 
> - [ ] Library: <Specify the library name>
> - [ ] Plug-in: <Specify the plugin name>
> - [ ] Engine
> - [ ] Documentation
> - [ ] Regression tests
> - [ ] Continous Integration (CI) scripts

> ### Impact of the pull request

> - [ ] Require a change on Quality of Results (QoR)
> - [ ] Break back-compatibility. If so, please list who may be influenced.

### CODE REVIEW

**Comment by KochynVolodymyr** [comment]
@skyler-rs it doesn't make sense to have it for one task since it will be from 0 to 100. Sure I can do it but I don't see the point.

**Comment by KochynVolodymyr** [comment]
image

File: src/Compiler/TaskManager.cpp [hunk 1]
```diff
@@ -127,16 +127,18 @@ void TaskManager::startAll() {
   m_runStack.append(m_tasks[TIMING_SIGN_OFF]);
   m_runStack.append(m_tasks[POWER]);
   m_runStack.append(m_tasks[BITSTREAM]);
-  run();
+  m_taskCount = m_runStack.count();
   emit started();
+  run();
```
Comment (ravic-rs) [inline src/Compiler/TaskManager.cpp:132]: Why was this moved after "emit started()" ?

File: src/MainWindow/main_window.cpp [hunk 2]
```diff
@@ -195,6 +195,19 @@ void MainWindow::createMenus() {
 
   helpMenu = menuBar()->addMenu("&Help");
   helpMenu->addAction(aboutAction);
+
+  m_progressWidget = new QWidget;
```
Comment (ravic-rs) [inline src/MainWindow/main_window.cpp:199]: Any reason to not create a separate function for creating this wedget ? e.g., createProgressWidget()

File: src/Compiler/TaskManager.cpp (hunk 1 above)
Comment (KochynVolodymyr) [reply src/Compiler/TaskManager.cpp:132]: because run() will not exit until task is running

File: src/MainWindow/main_window.cpp (hunk 2 above)
Comment (KochynVolodymyr) [reply src/MainWindow/main_window.cpp]: Done

File: src/Compiler/TaskManager.cpp (hunk 1 above)
Comment (ravic-rs) [reply src/Compiler/TaskManager.cpp:132]: I didn't follow why run() was before emit started() before this PR. It seems like we are changing a behavior which I am trying to understand.

Comment (KochynVolodymyr) [reply src/Compiler/TaskManager.cpp:132]: Before this change there was nothing for started() signal. Nothing was connected. No any functionality change

Comment (ravic-rs) [reply src/Compiler/TaskManager.cpp:132]: Thanks, I see the connection now. 
https://github.com/os-fpga/FOEDAG/pull/554/files#diff-9e31125b4e2b6bd449e09a6b478450cb54bf4fb9ddb23baadb06ad5d9a9bb149R402

File: src/Compiler/TaskManager.h [hunk 3]
```diff
@@ -109,6 +114,7 @@ class TaskManager : public QObject {
   QMap<uint, Task *> m_tasks;
   QVector<Task *> m_runStack;
   QMap<Task *, QVector<Task *>> m_rollBack;
+  std::optional<int> m_taskCount{std::nullopt};
```
Comment (ravic-rs) [inline src/Compiler/TaskManager.h:117]: Can we use anything from Qt here for optional value? I see that this is likely the only std we are using in this primarily Qt file.

Comment (KochynVolodymyr) [reply src/Compiler/TaskManager.h:117]: I don't see the problem with using anything from the std. This is even better.

Comment (KochynVolodymyr) [reply src/Compiler/TaskManager.h:117]: This is the question for discussion but I think we should use std there it is possible.

Comment (ravic-rs) [reply src/Compiler/TaskManager.h:117]: I am looking to see if we can separate out Qt vs Std code where we can for testing reasons down the line.

**Review by ravic-rs — APPROVED** [review]
Looks good.
